workers/unity/Packages/io.improbable.worker.sdk/IOStream.cs

```diff
@@ -0,0 +1,156 @@
+            }
+
+            var bytesWritten = 0L;
+            fixed (byte* dataToWrite = data)
+            {
+                bytesWritten = CIO.StreamWrite(stream, dataToWrite, 1);
+            }
+
+            if (bytesWritten != -1)
+            {
+                return bytesWritten;
+            }
+
+            var rawError = CIO.StreamGetLastError(stream);
+            throw new IOException(ApiInterop.FromUtf8Cstr(rawError));
+        }
+
+        public long Read(uint bytesToRead, out byte[] streamData)
+        {
+            CheckIfStreamClosed();
+
+            streamData = new byte[bytesToRead];
+
+            var bytesRead = 0L;
+            fixed (byte* streamDataPointer = streamData)
+            {
+                bytesRead = CIO.StreamRead(stream, streamDataPointer, bytesToRead);
+            }
+
+            if (bytesRead != -1)
+            {
+                return bytesRead;
+            }
+
+            var rawError = CIO.StreamGetLastError(stream);
+            throw new IOException(ApiInterop.FromUtf8Cstr(rawError));
+        }
+
+        public long Read(byte[] streamData)
+        {
+            CheckIfStreamClosed();
+
+            var bytesToRead = (uint) streamData.Length;
+            var bytesRead = 0L;
+            fixed (byte* streamDataPointer = streamData)
+            {
+                bytesRead = CIO.StreamRead(stream, streamDataPointer, bytesToRead);
+            }
+
+            if (bytesRead != -1)
+            {
+                return bytesRead;
+            }
+
+            var rawError = CIO.StreamGetLastError(stream);
+            throw new IOException(ApiInterop.FromUtf8Cstr(rawError));
+        }
+
+        public long Peek(uint bytesToPeek, out byte[] streamData)
+        {
+            CheckIfStreamClosed();
+
+            streamData = new byte[bytesToPeek];
+
+            var bytesPeeked = 0L;
+            fixed (byte* streamDataPointer = streamData)
+            {
+                bytesPeeked = CIO.StreamPeek(stream, streamDataPointer, bytesToPeek);
+            }
+
+            if (bytesPeeked != -1)
+            {
+                return bytesPeeked;
+            }
+
+            var rawError = CIO.StreamGetLastError(stream);
+            throw new IOException(ApiInterop.FromUtf8Cstr(rawError));
+        }
+
+        public long Ignore(uint bytesToIgnore)
+        {
+            CheckIfStreamClosed();
+
+            var bytesIgnored = CIO.StreamIgnore(stream, bytesToIgnore);
+
+            if (bytesIgnored != -1)
+            {
+                return bytesIgnored;
+            }
+
+            var rawError = CIO.StreamGetLastError(stream);
+            throw new IOException(ApiInterop.FromUtf8Cstr(rawError));
+        }
+
+        public uint GetRemainingCapacity()
+        {
+            return CIO.StreamGetRemainingWriteCapacityBytes(stream);
+        }
+
+        private void CheckIfStreamClosed()
```

Comment:
Nit: Prefer `ThrowIfStreamClosed` over `CheckIfStreamClosed`, I would assume `CheckIfStreamClosed` would return me a result rather than throwing an exception.